src/misc.rs

```diff
@@ -316,3 +317,68 @@ impl LateLintPass for PatternPass {
         }
     }
 }
+
+
+/// **What it does:** This lint checks for the use of bindings with a single leading underscore
+///
+/// **Why is this bad?** A single leading underscore is usually used to indicate that a binding
+/// will not be used. Using such a binding breaks this expectation.
+///
+/// **Known problems:** This lint's idea of a "used" variable is not quite the same as in the
+/// built-in `unused_variables` lint. For example, in the following code
+/// ```
+/// fn foo(_y: u32) -> u32) {
+///     let _x = 1;
+///     _x +=1;
+///     y
+/// }
+/// ```
+/// _x will trigger both the `unused_variables` lint and the `used_underscore_binding` lint.
+///
+/// **Example**:
+/// ```
+/// let _x = 0;
+/// let y = _x + 1; // Here we are using `_x`, even though it has a leading underscore.
+///                 // We should rename `_x` to `x`
+/// ```
+declare_lint!(pub USED_UNDERSCORE_BINDING, Warn,
+              "using a binding which is prefixed with an underscore");
+
+#[derive(Copy, Clone)]
+pub struct UsedUnderscoreBinding;
+
+impl LintPass for UsedUnderscoreBinding {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(USED_UNDERSCORE_BINDING)
+    }
+}
+
+impl LateLintPass for UsedUnderscoreBinding {
+    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        let needs_lint = match expr.node {
+            ExprPath(_, ref path) => {
+                let ident = path.segments.last()
+                                .expect("path should always have at least one segment")
+                                .identifier;
+                ident.name.as_str().chars().next() == Some('_') //starts with '_'
+                && ident.name.as_str().chars().skip(1).next() != Some('_') //doesn't start with "__"
+                && ident.name != ident.unhygienic_name //not in macro
+                && cx.tcx.def_map.borrow().values().any(|res| match res.base_def {
```

Comment:
uhm... what's the intended purpose of this statement? If I'm reading this right you are checking whether the `def_map` contains any `DefLocal`s... I think that's true for any non-trivial program...